src/vscode-dts/vscode.proposed.terminalShellIntegration.d.ts

```diff
@@ -0,0 +1,207 @@
+
+	export interface TerminalShellExecution {
+		// TODO: This is circular, is that fine? Should we add a `TerminalShellExecutionEvent`?
+		/**
+		 * The {@link Terminal} the command was executed in.
+		 */
+		terminal: Terminal;
+
+		/**
+		 * The full command line that was executed, including both the command and arguments.
+		 * The accuracy of this value depends on the shell integration implementation:
+		 *
+		 * - It may be undefined or the empty string until {@link onDidEndTerminalShellExecution} is
+		 *   fired.
+		 * - It may be inaccurate initially if the command line is pulled from the buffer directly
+		 *   via the [`OSC 633/133 ; A`, `B` and `C` sequences](https://code.visualstudio.com/docs/terminal/shell-integration#_vs-code-custom-sequences-osc-633-st).
+		 * - It may contain line continuation characters and/or parts of the right prompt.
+		 * - It may be inaccurate if the shell integration does not support command line reporting
+		 *   via the [`OSC 633 ; E` sequence](https://code.visualstudio.com/docs/terminal/shell-integration#_vs-code-custom-sequences-osc-633-st).
+		 */
+		commandLine: string | undefined;
+
+		/**
+		 * The working directory that was reported by the shell when this command executed. This
+		 * will be a {@link Uri} if the string reported by the shell can reliably be mapped to the
+		 * connected machine. This requires the shell integration to support working directory
+		 * reporting via the [`OSC 633 ; P`](https://code.visualstudio.com/docs/terminal/shell-integration#_vs-code-custom-sequences-osc-633-st)
+		 * or `OSC 1337 ; CurrentDir=<Cwd> ST` sequences.
+		 */
+		cwd: Uri | string | undefined;
+
+		/**
+		 * The exit code reported by the shell.
+		 */
+		exitCode: Thenable<number | undefined>;
+
+		/**
+		 * Creates a stream of raw data (including escape sequences) that is written to the
+		 * terminal. This will only include data that was written after `stream` was called for the
+		 * first time, ie. you must call `dataStream` immediately after the command is executed via
+		 * {@link executeCommand} or {@link onDidStartTerminalShellExecution} to not miss any data.
+		 *
+		 * @example
+		 * // Log all data written to the terminal for a command
+		 * const command = term.shellIntegration.executeCommand({ commandLine: 'echo "Hello world"' });
+		 * const stream = command.createDataStream();
+		 * for await (const data of stream) {
+		 *   console.log(data);
+		 * }
+		 */
+		createDataStream(): AsyncIterable<string>;
+	}
+
+	export interface Terminal {
+		/**
+		 * An object that contains [shell integration](https://code.visualstudio.com/docs/terminal/shell-integration)-powered
+		 * features for the terminal. This will always be undefined immediately after the terminal
+		 * is created. Listen to {@link window.onDidActivateTerminalShellIntegration} to be notified
+		 * when shell integration is activated for a terminal.
+		 */
+		shellIntegration: TerminalShellIntegration | undefined;
+	}
```

Comment:
Should probably include a note that shell integration may _never_ activate for a terminal.